### PR TITLE
NP-6624 - DateGroupingExpr and DateTimeGroupingExpr

### PR DESCRIPTION
**Pull request by tala-abuadas**
https://nanopay.atlassian.net/browse/NP-6624

### CODE REVIEW

File: src/foam/mlang/expr/DateTimeGroupingExpr.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    'foam.nanos.auth.CreatedAware',
+    'java.util.Arrays',
+    'java.util.Comparator',
+    'java.util.Date'
+  ],
+
+  properties: [
+    {
+      class: 'Enum',
+      of: 'foam.mlang.expr.DateTypes',
```
Comment (jlhughes) [inline src/foam/mlang/expr/DateTimeGroupingExpr.js]: @lchanmann 
Another duplication 
foam.mlang.expr.DateTypes and 
foam.time.TimeUnit 
What dependencies do we have on expr.DateTypes?

File: src/foam/mlang/expr/DateGroupingExpr.js [hunk 2]
```diff
@@ -12,91 +12,69 @@ foam.CLASS({
...
-          if ( 
-            objDiffFromTodayConverted >= group.low && 
-            objDiffFromTodayConverted < group.high 
-          ) groupName = group.name;
-        })
-
-        return groupName;
+        if ( obj.created ) {
+          var dateCreated = new Date(obj.created);
+          if ( this.dateGroupingType == 'DAYS' ) {
```
Comment (tharmaman) [inline src/foam/mlang/expr/DateGroupingExpr.js]: This should be comparing to the enum and not the string value like `foam.time.TimeUnit.DAY`

File: src/foam/mlang/expr/DateGroupingExpr.js [hunk 3]
```diff
@@ -12,91 +12,69 @@ foam.CLASS({
...
-
-        for ( int i = 0; i < dateGroupsSorted.length; i++ ){
-          DateGrouping group = dateGroupsSorted[i];
-
-          if (
-            objDiffFromTodayConverted >= group.getLow() &&
-            objDiffFromTodayConverted < group.getHigh()
-          ) {
-            return group.getName();
+          if ( getDateGroupingType().getName().equals("DAYS") ) {
```
Comment (tharmaman) [inline src/foam/mlang/expr/DateGroupingExpr.js]: Same here compare the enum and not the string value

File: src/foam/mlang/expr/DateGroupingExpr.js [hunk 4]
```diff
@@ -12,91 +12,69 @@ foam.CLASS({
...
-          ) groupName = group.name;
-        })
-
-        return groupName;
+        if ( obj.created ) {
+          var dateCreated = new Date(obj.created);
+          if ( this.dateGroupingType == 'DAYS' ) {
+            return dateCreated.toLocaleString([], { day: 'numeric', month: 'long',  year: 'numeric' });
+          }
+          else if ( this.dateGroupingType == 'MONTHS' ) {
```
Comment (tharmaman) [inline src/foam/mlang/expr/DateGroupingExpr.js]: same as above

File: src/foam/mlang/expr/DateGroupingExpr.js [hunk 5]
```diff
@@ -12,91 +12,69 @@ foam.CLASS({
...
+            return dateCreated.toLocaleString([], { month: 'long',  year: 'numeric' });
+          }
+          else if ( this.dateGroupingType == 'YEARS' ) {
```
Comment (tharmaman) [inline src/foam/mlang/expr/DateGroupingExpr.js]: same as above

File: src/foam/mlang/expr/DateGroupingExpr.js [hunk 6]
```diff
@@ -12,91 +12,69 @@ foam.CLASS({
...
-
-          if (
-            objDiffFromTodayConverted >= group.getLow() &&
-            objDiffFromTodayConverted < group.getHigh()
-          ) {
-            return group.getName();
+          if ( getDateGroupingType().getName().equals("DAYS") ) {
+            return Integer.toString(c.get(Calendar.DAY_OF_MONTH)) + " " + c.getDisplayName(Calendar.MONTH, Calendar.LONG, new Locale("en") ) + ", " + Integer.toString(c.get(Calendar.YEAR));
+          }
+          else if ( getDateGroupingType().getName().equals("MONTHS") ) {
```
Comment (tharmaman) [inline src/foam/mlang/expr/DateGroupingExpr.js]: same as above

File: src/foam/mlang/expr/DateGroupingExpr.js [hunk 7]
```diff
@@ -12,91 +12,69 @@ foam.CLASS({
...
+            return c.getDisplayName(Calendar.MONTH, Calendar.LONG, new Locale("en") ) + " " + Integer.toString(c.get(Calendar.YEAR));
+          }
+          else if ( getDateGroupingType().getName().equals("YEARS") ) {
```
Comment (tharmaman) [inline src/foam/mlang/expr/DateGroupingExpr.js]: same as above

File: src/foam/mlang/expr/DateTimeGroupingExpr.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
```
Comment (tharmaman) [inline src/foam/mlang/expr/DateTimeGroupingExpr.js]: 2022

**Review by tharmaman — CHANGES_REQUESTED** [review]

**Review by tharmaman — APPROVED** [review]

**Review by jlhughes — APPROVED** [review]

**Pull request merged**